cert-manager incubation proposal (#2)
SgtCoDFish/toc: proposals/incubation/cert-manager.md

# cert-manager Incubation Proposal

cert-manager joined the CNCF as a sandbox project just over a year ago, and we firmly believe that the project is ready to take the next step and graduate to the Incubation stage.

cert-manager provides reliable, automated Kubernetes-native management of X.509 certificates. It reduces the risk of certificate-related outages and enables PKI at scale, and is treated by many teams as de-facto part of the core components of a Kubernetes or OpenShift deployment.

> The project comprises cert-manager itself - which adds Kubernetes Custom Resource Definitions (CRDs) for certificate-related concepts, and provides controllers to manage them - along with a selection of subprojects which focus on solving other issues relating to certificate management.

Reviewer: > The project comprises cert-manager itself.. along with a selection of subprojects which focus on solving other issues relating to certificate management

This makes it a little unclear whether it'd be just cert-manager/cert-manager that'd be graduating or the other subprojects too. It'd maybe be good to state clearly that just cert-manager/cert-manager must meet the incubation requirements?

Author: This is a thought-provoking comment. For sure the main thing here is cert-manager itself, but is it "cert-manager the product" or "cert-manager the project" that's being proposed for incubation?

E.g. [Argo's CNCF project page](https://www.cncf.io/projects/argo/) has a GitHub link directly to [the main workflows repo](https://github.com/argoproj/argo-workflows) but the argo website lists all of the sub projects (e.g. argo-cd, events, etc) in the title bar. Is argo-cd an incubating project? If they started a new greenfield project called "argo-certs" which managed TLS certs, would that qualify as incubating? I don't know how the CNCF draws those lines.

Maybe something like this? :thinking: 

```suggestion
The project comprises cert-manager itself - which adds Kubernetes Custom Resource Definitions (CRDs) for certificate-related concepts, and provides controllers to manage them - along with a selection of subprojects which focus on solving other issues relating to certificate management. We'll primarily discuss cert-manager in this proposal, although we do intend to keep developing the other projects, all of which tie into cert-manager.
```

Reviewer: Interestingly Argo does explicitly mention that 'all projects must meet incubation requirements' https://github.com/cncf/toc/blob/main/proposals/incubation/argo.adoc#:~:text=As%20a%20CNCF%20incubation%20project%2C%20all%20top%20level%20projects%20must%20meet%20incubation%20maturity%20requirements.

> We'll primarily discuss cert-manager in this proposal, although we do intend to keep developing the other projects, all of which tie into cert-manager

I tend to read the incubation requirements in a way that to me it would seem that if we where to include a bunch of tools into the proposal, then each one of them should meet the requirements separately, which currently may not be the case I think. So, I'd be looking for a clarification for whether a) all tools mentioned go under one project and one of the tools meeting the requirements is enough or b) there are associated tools, but we don't propose those to incubation, please only consider cert-manager repo/tool

Author: Oh, I'd missed that part in argo's proposal, thanks for linking. I don't think any of the subprojects would pass the incubation criteria individually, and so I lean towards (b) in what you described:

>  b) there are associated tools, but we don't propose those to incubation, please only consider cert-manager repo/tool

I'll make that change :+1:

The main cert-manager repo has 8.4k stars on GitHub and hundreds of projects downstream which depend on cert-manager code in some way. Thousands of people have contributed to the project in some way and cert-manager's use and prominence remains strong.

## Incubation Criteria Checklist

> Based on the [graduation criteria][GraduationCriteria], cert-manager seems to clearly tick the boxes required for incubation:

Reviewer 2: There's one remaining item on that list: 
> Specifications must have at least one public reference implementation

But that could be the `approver-policy` implementation of [policy.cert-manager.io](https://github.com/cert-manager/approver-policy/blob/main/docs/api/api.md): 

```
apiVersion: policy.cert-manager.io/v1alpha1
kind: CertificateRequestPolicy
```

Or sample-external-issuer of the [external issuer protocol](https://cert-manager.io/docs/contributing/external-issuers/).

Or the [sample webhook implementation](https://github.com/cert-manager/webhook-example) of the DNS01 webhook protocol

Author: I interpreted that "Specifications" section as not being relevant to us, which was why I left it out.

I felt like it was more for projects like [CNI](https://github.com/containernetworking/cni) for whom the spec is a primary goal of the project. In my view, cert-manager doesn't provide a spec; rather, it provides APIs which people can interact with.

### Production Use

> Document that it is being used successfully in production by at least three independent end users

> cert-manager is used in production by _many_ more than three independent end users. Several major cloud native projects including [Istio][IstioCertManager], the [Kubernetes Cluster API][ClusterAPICertManager] and [Knative][KnativeCertManager] either suggest or require that cert-manager be installed.

Reviewer 3: fwiw, ArgoCD uses cert-manager for its own demo instance: https://cd.apps.argoproj.io/applications/cert-manager

(We certainly use cert-manager for our k8s clusters, we've occasionally made PRs to list ourselves as a consumer of projects, but haven't generally. I'm not personally sure how I feel about such links...)

Author: Thanks for the heads up! I'll add the argo link, great to have another CNCF example.

If you end up wanting to raise a PR for users.md please do feel free :grin:

> Walmart uses cert-manager too, and mentioned it in a [Keynote speech they gave][WalmartKeynote].

Reviewer 4: We could add https://www.youtube.com/watch?v=XlkGsBGk9W4 as well. I think that clearly shows that something of adequate quality and scope.

Author: Thanks, added!

According to the [companies-table DevStats dashboard][CompaniesTable] over a thousand different companies (plus ~1330 "Independents") have made some kind of contribution (defined as somebody who made a review, comment, commit, created PR or issue) to cert-manager. That implies that the floor of cert-manager's reach can be measured in the thousands, but there could be many more users who happily use cert-manager without contributing back or raising GitHub issues.

One prominent example of a cert-manager user might be the CNCF themselves, using cert-manager [in the DevStats project][DevStatsCertManager]!

In addition we might measure some production use through the many large companies who have seen fit to create addons for cert-manager to help it better integrate into their environments. That includes (at the time of writing):

- at least [49 public repositories][ExternalDNSSolvers] implementing an external ACME DNS-01 solver
- at least [8 public repositories][ExternalIssuers] implementing complete external issuers for certificates

> Finally, we track a few users in a [USERS.md file][USERSmd] on our repository, although that file doesn't fully represent the scale of cert-manager's prominence in clusters today.

Reviewer 4: How about 

```suggestion
Finally, we offer a place for users to identify themselves in a [USERS.md file][USERSmd] on our repository.
```

As it's not so much us "tracking" users, but this also makes it clearer it is self-selecting.

I don't think we need to be too defensive here. The criteria say:

> Document that it is being used successfully in production by *at least three* independent end users

We've emphasised in the rest of the section that we think there are a lot of users, and the reader will understand that it isn't a complete list. It also shows three independent end users, which is the bar that is being set, so we don't need to stress that we have a lot more in order for them to be satisfied with the impact.

Author: Good shout, I'll adopt your language here. Fully agreed.

> 

Reviewer 2: optional: We could list a couple of "production users" listed on the ArtifactHub page:
![image](https://user-images.githubusercontent.com/978965/154064685-c0a6aa3b-2a36-4915-a4c2-4f0002bf84c0.png)
https://artifacthub.io/packages/helm/cert-manager/cert-manager

Author: I like the idea and I'd not thought to look on ArtifactHub. That said, I think I'd prefer to not add these to the doc; we have plenty of orgs listed currently (and we only needed 3).

### Committer Numbers

> Have a healthy number of committers

cert-manager currently has 9 maintainers across 3 organisations. Community contributions from many different organizations are very common and gladly received.

The [companies-table DevStats dashboard][CompaniesTable] is illustrative. We've seen:

- 4 different companies which have made over a thousand contributions each,
- 8 different companies which have made over a hundred contributions each,
- and over a thousand different companies which have contributed something to the project.

> In 2021, cert-manager saw [102 new first-time-contributors][NewContributors2021] to the project. Drilling down, cert-manager v1.7 saw at least 7 different individual community contributors, and one of the major features in that release was [contributed by a community member][ReleaseNotes17].

Reviewer 5: > Drilling down

What do we mean by that?

```suggestion
In 2021, cert-manager saw [102 new first-time-contributors][NewContributors2021] to the project. Our latest release as of creating this PR, cert-manager v1.7, saw at least 7 different individual community contributors, and one of the major features in this release was [contributed by a community member][ReleaseNotes17].
```

Author: I meant "to focus on a specific release as an example", but that wasn't clear and your wording is better; I'll use your wording!

2022 has started strongly too; at the time of writing there have been [21 new contributors][NewContributors2022] already in this calendar year.

In terms of code changes only and ignoring other types of contribution, we see an [average of 10 different individual committers][CommittersCertManager] per month in the 6 months before February 2022 for cert-manager alone, rising to an [average of 20 per month][CommittersAllRepos] across all repositories in the same time period.

### Ongoing Commits

> Demonstrate a substantial ongoing flow of commits and merged contributions

> All recent releases of cert-manager have included major features contributed by both the core maintainer team and community members, driven by a contiuous flow of commits every month.

Reviewer 5: ```suggestion
All recent releases of cert-manager have included major features contributed by both the core maintainer team and community members, driven by a continuous flow of commits every month.
```

Looking at CNCF DevStats, cert-manager saw an average of [39 pull requests opened][AvgOpenedPRsCertManager] per month in the 6 months before February 2022 (excluding bots), with an average of [30 PRs merged per month][AvgMergedPRsCertManager] in the same time period.

> The above numbers only count the main cert-manager repo; as mentioned previously, there are several cert-manager subprojects being regularly worked on too, including the cert-manager website.
> 
> Across all repos we see an average of [78 merged PRs][AvgMergedPRsAllRepos] per month in the same time period as above.

Reviewer 4: ```suggestion
The above numbers only count the main cert-manager repo; as mentioned previously, there are several cert-manager subprojects being regularly worked on too, including the cert-manager website. Across all repos we see an average of [78 merged PRs][AvgMergedPRsAllRepos] per month in the same time period as above.
```

These two points seem to be directly tied together, so I would suggest putting them in the same paragraph.

### Versioning

> A clear versioning scheme

> cert-manager has had a clear versioning scheme and a regular release cadence for over a year. Details can be seen on the cert-manager website under our [Supported Releases][] page, which documents full and clear support lifecycles for the main cert-manager project. We broadly use SemVer.

Reviewer 3: This link is broken

> cert-manager has had a clear versioning scheme and a regular release cadence for over a year. Details can be seen on the cert-manager website under our [Supported Releases][] page, which documents full and clear support lifecycles for the main cert-manager project. We broadly use SemVer.

Reviewer 5: > We broadly use SemVer.

Is this really true?

We've introduced breaking changes for minor versions

Author: That's fair - I think SemVer is closest to what we do, but probably it's not helpful to actually call it out by name given that we do deviate to some extent. I'll remove the ref :+1:

Author: I'm changing it to say we're similar to k8s - avoid breaking changes, but if we make one then we try to signpost far in advance.

> Being closely tied to Kubernetes, cert-manager also makes great efforts to be clear which versions of Kubernetes itself are supported by a given version of cert-manager.

Reviewer 5: ```suggestion
Being closely tied to Kubernetes, cert-manager also makes great efforts to be clear which versions of Kubernetes it is compatible with.
```

### Security Process

> Clearly documented security processes explaining how to report security issues

> A security reporting mechanism was added in 2021 with a clear [security process][] and a dedicated email address: `cert-manager-security@googlegroups.com`. This security policy applies to all subprojects under the cert-manager umbrella too.

Reviewer 5: Missing a link here, do we mean https://github.com/cert-manager/cert-manager/blob/master/SECURITY.md?

Author: The link is actually defined here, with the same name as the link text: https://github.com/SgtCoDFish/toc/pull/2/files#diff-8750b84c65e4361ee6d6629a7906376c7e833ee55afddbf434657719180d2065R133

It does link to that file.

> 

Reviewer 5: We are missing this part from the security criteria: 

> and describing how the project provides updated releases or patches to resolve security vulnerabilities

We could write a few sentences on our process for backporting and releasing patch versions?

Author: I think I'd just kinda assumed they'd read the SECURITY.md file, but I think it would totally be worthwhile to write a little in this doc too. Will add.

## The cert-manager Project

> ### Encouraging Open Source

Reviewer: Suggestion: mention that we also interact with community on Slack and that we mark 'good first issues' to encourage contributions

Author: Good shout, thanks :+1:

# cert-manager Incubation Proposal

cert-manager joined the CNCF as a sandbox project just over a year ago, and we firmly believe that the project is ready to take the next step and graduate to the Incubation stage.

cert-manager provides reliable, automated Kubernetes-native management of X.509 certificates. It reduces the risk of certificate-related outages and enables PKI at scale, and is treated by many teams as de-facto part of the core components of a Kubernetes or OpenShift deployment.

The project comprises cert-manager itself - which adds Kubernetes Custom Resource Definitions (CRDs) for certificate-related concepts, and provides controllers to manage them - along with a selection of subprojects which focus on solving other issues relating to certificate management.

> The main cert-manager repo has 8.4k stars on GitHub and hundreds of projects downstream which depend on cert-manager code in some way. There are thousands of companies which have contributed in some way and cert-manager's use and prominence remains strong.

Reviewer 4: > thousands of companies which have contributed in some way

Do we want to call out companies, or just people? "thousands of people which have contributed"?

Author: I moved this bit from "production use" where I was talking more about companies. Definitely would read better as "people", will change :+1:

## Incubation Criteria Checklist

Based on the [graduation criteria](https://github.com/cncf/toc/blob/da2c6f1a14ea9592ca0d1794da376e0c874b5a6f/process/graduation_criteria.adoc#incubating-stage), cert-manager seems to clearly tick the boxes required for incubation:

> ### Production Use

Reviewer 4: You could point to https://github.com/cert-manager/cert-manager/blob/master/USERS.md, there's not many listed there, but there are 3 non-Jetstack companies listed.

There is this presentation from Walmart where they talk about using cert-manager: https://youtu.be/sfPFrvDvdlk?t=709

I think these are worth including as well, as they are pointing to production use by companies, whereas most of the things listed here are inferring production use from the fact that companies are contributing.

Author: I was a little squeamish about USERS.md just because it's so much smaller than our actual userbase. It doesn't feel representative of the actual usage... although, it's _something_. I'll add it in!

The walmart presentation is a great shout, thanks :+1:

- at least [8 public repositories](https://cert-manager.io/docs/configuration/external/) implementing complete external issuers for certificates

### Committer Numbers

> Have a healthy number of committers

cert-manager currently has 9 maintainers across 3 organisations. Community contributions from many different organizations are very common and gladly received.

The [companies-table DevStats dashboard](https://certmanager.devstats.cncf.io/d/5/companies-table?orgId=1) is illustrative. We've seen:

- 4 different companies which have made over a thousand contributions each,
- 8 different companies which have made over a hundred contributions each,
- and over a thousand different companies which have contributed something to the project.

In 2021, cert-manager saw roughly 45 new first-time-contributors to the codebase. cert-manager v1.7 saw at least 7 different individual community contributors, and one of the major features in that release was [contributed by a community member](https://cert-manager.io/docs/release-notes/release-notes-1.7/#additional-certificate-output-formats).

> In the last 30 days at the time of writing, there were [15 different first-time contributors to the project](https://certmanager.devstats.cncf.io/d/52/new-contributors-table?orgId=1&from=now-30d&to=now).

Reviewer 4: Can we say something more about the number of committers as well? The criteria is to have a healthy number of committers, rather than the number of companies.

Author: Works out at an average of 10 different committers per month over the last 6 months (and across most timescales, really); I've added some details :+1:

### Ongoing Commits

> > Demonstrate a substantial ongoing flow of commits and merged contributions

Reviewer 4: I suggest adding a count like average number of PRs merged to cert-manager projects per month for the last 6 months. Currently this section reads as though we are dodging the question a bit, when I think that we can directly show that we have a substantial ongoing flow of commits and merged contributions.

Author: Fair point, I'm adding a few relevant stats to improve this section :+1:
